staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go

```diff
@@ -143,107 +137,168 @@ func (h *peerProxyHandler) WrapHandler(handler http.Handler) http.Handler {
+			klog.Errorf("gvr %v is not served by anything in this cluster", gvr)
+			handler.ServeHTTP(w, r)
+			return
+		}
+
+		peerEndpoints, err := h.resolveServingLocation(peerServerIDs)
 		if err != nil {
 			gv := schema.GroupVersion{Group: gvr.Group, Version: gvr.Version}
 			klog.ErrorS(err, "error finding serviceable-by apiservers for the requested resource", "gvr", gvr)
 			responsewriters.ErrorNegotiated(apierrors.NewServiceUnavailable("Error getting ip and port info of the remote server while proxying"), h.serializer, gv, w, r)
 			return
 		}
 
-		// pass request to the next handler if found the gvr locally.
-		// TODO: maintain locally serviceable GVRs somewhere so that we dont have to
-		// consult the storageversion-informed map for those
-		if locallyServiceable {
-			handler.ServeHTTP(w, r)
-			return
+		rand := rand.Intn(len(peerEndpoints))
+		peerEndpoint := peerEndpoints[rand]
+		h.proxyRequestToDestinationAPIServer(r, w, peerEndpoint)
+	})
+}
+
+func (h *peerProxyHandler) startLocalDiscoveryCacheInvalidation(stopCh <-chan struct{}) {
+	go func() {
+		klog.Info("localDiscoveryCacheInvalidation goroutine started")
+		for {
+			select {
+			case <-h.localDiscoveryCacheTicker.C:
+				klog.V(4).Infof("Invalidating local discovery cache")
+				if err := h.populateLocalDiscoveryCache(); err != nil {
+					klog.Errorf("Failed to repopulate local discovery cache: %v", err)
+				}
+			case <-stopCh:
+				klog.Info("localDiscoveryCacheInvalidation goroutine received stop signal")
+				if h.localDiscoveryCacheTicker != nil {
+					h.localDiscoveryCacheTicker.Stop()
+					klog.Info("localDiscoveryCacheTicker stopped")
+				}
+				klog.Info("localDiscoveryCacheInvalidation goroutine exiting")
+				return
+			}
 		}
+	}()
+}
 
-		if len(peerEndpoints) == 0 {
-			klog.Errorf("gvr %v is not served by anything in this cluster", gvr)
-			handler.ServeHTTP(w, r)
-			return
+func (h *peerProxyHandler) populateLocalDiscoveryCache() error {
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(h.loopbackClientConfig)
+	if err != nil {
+		return fmt.Errorf("error creating discovery client: %w", err)
+	}
+
+	_, resourcesByGV, _, err := discoveryClient.GroupsAndMaybeResources()
+	if err != nil {
+		return fmt.Errorf("error getting API group resources from discovery: %w", err)
+	}
+
+	freshLocalDiscoveryResponse := map[schema.GroupVersion][]string{}
+	for gv, resources := range resourcesByGV {
+		if gv.Group == "" {
+			gv.Group = "core"
 		}
+		freshLocalDiscoveryResponse[gv] = []string{}
+		for _, resource := range resources.APIResources {
+			freshLocalDiscoveryResponse[gv] = append(freshLocalDiscoveryResponse[gv], resource.Name)
+		}
+	}
 
-		// otherwise, randomly select an apiserver and proxy request to it
-		rand := rand.Intn(len(peerEndpoints))
-		destServerHostPort := peerEndpoints[rand]
-		h.proxyRequestToDestinationAPIServer(r, w, destServerHostPort)
-	})
+	h.localDiscoveryResponseCache = freshLocalDiscoveryResponse
```

Comment:
Is this thread safe? We're still performing read/write operations on the map pointer without a lock, do we need an atomic pointer in this scenario?

Reply:
Protected with a write lock and also made it a atomic.Value type.